pallets/xcm-asset-manager/src/lib.rs

```diff
@@ -0,0 +1,404 @@
+			AssetIdType::<T>::insert(&asset_id, &asset_type);
+			AssetTypeId::<T>::insert(&asset_type, &asset_id);
+
+			Self::deposit_event(Event::<T>::ForeignAssetTypeRegistered { asset_id, asset_type });
+			Self::deposit_event(Event::<T>::ForeignAssetMetadataUpdated { asset_id, metadata });
+
+			Ok(())
+		}
+
+		// Update asset metadata
+		#[pallet::weight(T::WeightInfo::update_foreign_asset_metadata())]
+		pub fn update_foreign_asset_metadata(
+			origin: OriginFor<T>,
+			asset_id: T::AssetId,
+			metadata: AssetMetadata<BalanceOf<T>>,
+		) -> DispatchResult {
+			T::ForeignAssetModifierOrigin::ensure_origin(origin)?;
+
+			ensure!(AssetIdType::<T>::contains_key(&asset_id), Error::<T>::AssetIdDoesNotExist);
+			AssetIdMetadata::<T>::insert(&asset_id, &metadata);
+
+			Self::deposit_event(Event::<T>::ForeignAssetMetadataUpdated { asset_id, metadata });
+			Ok(())
+		}
+
+		/// Change the amount of units we are charging per execution second
+		/// for a given ForeignAssetType
+		/// 0 means not support
+		#[pallet::weight(T::WeightInfo::set_asset_units_per_second())]
+		pub fn set_asset_units_per_second(
+			origin: OriginFor<T>,
+			asset_id: T::AssetId,
+			units_per_second: u128,
+		) -> DispatchResult {
+			T::ForeignAssetModifierOrigin::ensure_origin(origin)?;
+
+			ensure!(AssetIdType::<T>::contains_key(&asset_id), Error::<T>::AssetIdDoesNotExist);
+
+			AssetIdUnitsPerSecond::<T>::insert(&asset_id, &units_per_second);
+
+			Self::deposit_event(Event::UnitsPerSecondChanged { asset_id, units_per_second });
+			Ok(())
+		}
+
+		/// Add the xcm type mapping for a existing assetId, other assetType still exists if any.
+		/// TODO: Change add_asset_type with internal function wrapper
+		#[pallet::weight(T::WeightInfo::add_asset_type())]
+		pub fn add_asset_type(
+			origin: OriginFor<T>,
+			asset_id: T::AssetId,
+			new_asset_type: T::ForeignAssetType,
+		) -> DispatchResult {
+			T::ForeignAssetModifierOrigin::ensure_origin(origin)?;
+
+			// TODO: TBD, mutate replace insert + ensure! as asset_id is always an update
+			ensure!(AssetIdType::<T>::contains_key(&asset_id), Error::<T>::AssetIdDoesNotExist);
+			ensure!(
+				!AssetTypeId::<T>::contains_key(&new_asset_type),
+				Error::<T>::AssetAlreadyExists
+			);
+
+			// Insert new asset type info
+			AssetIdType::<T>::insert(&asset_id, &new_asset_type);
+			AssetTypeId::<T>::insert(&new_asset_type, &asset_id);
+
+			Self::deposit_event(Event::ForeignAssetTypeRegistered {
+				asset_id,
+				asset_type: new_asset_type,
+			});
+			Ok(())
+		}
+
+		/// We do not allow the destroy of asset id so far; So at least one AssetTpye should be
+		/// assigned to existing AssetId Both asset_type and potential new_default_asset_type must
+		/// be an existing relation with asset_id
+		/// TODO: Change remove_asset_type with internal function wrapper
+		#[pallet::weight(T::WeightInfo::remove_asset_type())]
+		#[transactional]
+		pub fn remove_asset_type(
+			origin: OriginFor<T>,
+			asset_id: T::AssetId,
```

Comment:
I don't think we need the `asset_id` here, otherwise I can break the function by providing **existing but unmatching** `asset_id` and `asset_type`, which is incorrect but still passes the first two checks.

Reply:
agreed.

Reply:
Thanks, I saw the update.

Shall we also check the `new_default_asset_type` actually maps the same `asset_id` as the asset type to be removed? Right now it only checks existence.

Reply:
I have corrected the logic accordingly and add an unittest for it.